src/shadowbox/server/managed_access_key.ts

```diff
@@ -122,29 +122,14 @@ class ManagedAccessKeyRepository implements AccessKeyRepository {
   private nextId = 0;
   private NEW_USER_ENCRYPTION_METHOD = 'chacha20-ietf-poly1305';
   private reservedPorts: Set<number> = new Set();
+  private ssInstances = new Map<AccessKeyId, ShadowsocksInstance>();
 
   constructor(private proxyHostname: string, private configFile: AccessKeyConfigFile,
       private configJson: ConfigJson, private shadowsocksServer: ShadowsocksServer,
       private statsSocket: dgram.Socket, private stats: Stats) {
-    this.updateServer();
-  }
-
-  private updateServer(): Promise<void> {
-    const startInstancePromises = [];
     for (const accessKeyJson of this.configJson.accessKeys) {
-      startInstancePromises.push(
-          this.shadowsocksServer
-              .startInstance(
-                  accessKeyJson.port, accessKeyJson.password, this.statsSocket,
-                  accessKeyJson.encryptionMethod)
-              .then((ssInstance) => {
-                ssInstance.onInboundBytes(this.handleInboundBytes.bind(
-                    this, accessKeyJson.id, accessKeyJson.metricsId));
-              }));
+      this.startInstance(accessKeyJson);
```

Comment:
Good catch removing Promise-related stuff from the constructor. This will still fail fast though, i.e. if an access key cannot be started then we won't even try to start the remainder. Is that what you want?

Reply:
Do you mean it will fail fast because the Promise rejection won't be handled? I think we only get a warning. In any case, I added error handling to the startInstance method.

Reply:
Yes, because the rejection won't be handled. The change you made to `startInstance` is better for the constructor, but it's called in other places - generally better to have the client deal with the error, no?